### ISSUE TITLE
UserProfileState not saving after changing in a dialog called from SkillDialog

### ISSUE THREAD

**Issue by lshade** [microsoft/botframework-solutions#3339]
#### What project is affected?
Virtual Assistant

#### What language is this in?
C#

#### What happens?

In my Skills, I want to be able to forward the user's message, from the Skill back to the VA, if it doesn't answer a Skill prompt.
In the VA, I implemented a class that inherits the `SkillDialog` class, and in the `ContinueDialogAsync`, I check for the Skill sending the VA a `DialogArg`. That contains the utterance to send through the VA's MainDialog. 
Once the utterance reaches the MainDialog, it is routed to one of the VA's dialogs. In that VA Dialog, I change a property in the `UserProfileState`. All of this is working as expected.

The MainDialog is started again and hits the `IntroStepAsync`. The `UserProfileState` change from before is still in there.

However, when the user's next message comes in, I check the `UserProfileState` in `RouteStepAsync` and the change from before is gone. The rest of the `UserProfileState` is correct, just the change made from before is missing.

This issue only happens when I have no breakpoints, or when I am using Cosmos. If I use `services.AddSingleton<IStorage, MemoryStorage>();` instead, the state change is still there. This leads me to think there's some async issue. But if that same VA dialog with the `UserProfileState` change is called directly from MainDialog (and not originally from `SkillDialog`), then there's no issue.

I also set a breakpoint in the DefaultActivityHandler's `OnTurnAsync` to see if it does save the user state after each turn, and it does hit the breakpoint.

#### What are the steps to reproduce this issue?

TLDR. Start the MainDialog from inside the SkillDialog, then start a dialog that changes the UserProfileState with Cosmos, then restart the MainDialog. The change to UserProfileState is gone by the time the next message gets to `RouteStepAsync`

#### What were you expecting to happen?

For `UserProfileState` to remain the same between the MainDialog's `IntroStepAsync` and `RouteStepAsync`. For some reason the change is reverted in this specific case.

#### Any screenshots or additional context?

I'm available to screen share

**Comment by darrenj** [comment]
Hi - Can you share the code snipped where you manipulate user profile state please?

**Comment by lshade** [comment]
Yes, here it is:
```
            // Increment turns 
            var userProfile = await _accessor.GetAsync(stepContext.Context, () => new UserProfileState());
            userProfile.PersonalityChatTurnsBeforePrompt = ++userProfile.PersonalityChatTurnsBeforePrompt;
            await _accessor.SetAsync(stepContext.Context, userProfile, cancellationToken);
```

We also recently tried adding this directly after, but the issue persists:
```
            // Save it to state to resolve bug with change reverting
            await _userState.SaveChangesAsync(stepContext.Context, true, cancellationToken);
            await stepContext.Context.SendActivityAsync(new Activity(type: ActivityTypes.Trace, text: $"PersonalityChatTurnsBeforePrompt incremented and saved."));

**Comment by darrenj** [comment]
SaveChanges was what I was looking for so great you've already tried that.  I have a theory - can you trace out the contents of each Activity - i.e. the first one triggering the Intro step and the second one triggering the secondary step.

Looking to see if `From.Id` and `Conversation.Id` are the same..

**Comment by darrenj** [comment]
Any update @lshade ?

**Comment by lshade** [comment]
Yes, `From.Id` and `Conversation.Id` are the same in IntroStepAsync and RouteStepAsync.

Interestingly it works locally if we have one of the following:
1. using locally memory instead of Cosmos in Startup.cs
2. SaveChangesAsync right after setting the property of the user state
3. Trace activity added in IntroStepAsync and RouteStepAsync

Right now the bot is deployed with both 2 and 3, but it _doesn't_ work in the deployed bots. It seems like locally it's given enough time to change the state variable, but in the deployed environment it isn't setting.

**Comment by darrenj** [comment]
Hi Lisa - could you confirm the BF SDK version you are using? I **think** this might have been an issue that was discovered and fixed in 4.9.1 but would be good to confirm.

**Comment by lshade** [comment]
Ah we are using 4.7.2 still. Thanks Darren.

Tagging @radoslav84 who's taking this over for me. Looks like it could be resolved if you upgrade the bots to 4.9.1.

**Comment by darrenj** [comment]
OK - it would be good to update the packages to the latest version and see if you can repro. For what it's worth I have performed the same steps with the latest bits and can't repro the issue. Given how you describe it not happening with memory or locally it fits well with the race condition idea.

Feel free to re-open if you can repro the problem in the latest bits.
